### ISSUE TITLE
[suggestion] DSL for event filters

### ISSUE THREAD

**Issue by appetrosyan** [hyperledger/iroha#3068]
### Feature request

I propose designing a small Domain-specific language implemented as a procedural macro crate that generates code that produces the following:

Examples
```rust
let asset_filter: FilterBox = filter!(asset "rose##alice@wonderland" added);
let passthrough: FilterBox = filter!(*);
let domain_metadata: FilterBox = filter!(domain metadata["key"])
let account = filter!(asset "rose##alice@wonderland" metadata_inserted);
```

### Motivation

Composing event filters is extremely cumbersome, because the entire hierarchy of the nets must be remembered by the person writing the filter. 

### Who can help?

@Arjentix

**Comment by DCNick3** [comment]
I've decided to make the DSL not macro-based, but based on builder pattern. For this I would also like to refactor the events and event filters first to be more orthogonal

**Comment by DCNick3** [comment]
This is mostly addressed with the merge of #4240. Some improvements (#4259, #4356) will follow
